### PR TITLE
Added diagnostic/suggestion for core::pin::pin! in addition to Box::pin if Unpin isn't implemented

### PR DESCRIPTION
**Pull request by ndrewxie**
Currently, if an operation on `Pin<T>` is performed that requires `T` to implement `Unpin`, the diagnostic suggestion is to use `Box::pin` ("note: consider using `Box::pin`").

As others have noted, using `pin!` avoids a heap allocation, so it should be suggested as well (and probably over `Box::pin` whenever possible).

This PR adds the diagnostic "consider using `Box::pin` if you want ownership of the pinned value", and updates tests to match.

Addresses/attempts to address #109964

### CODE REVIEW

File: library/core/src/marker.rs [hunk 1]
```diff
@@ -823,7 +823,7 @@ unsafe impl<T: ?Sized> Freeze for &mut T {}
 /// [`pin` module]: crate::pin
 #[stable(feature = "pin", since = "1.33.0")]
 #[rustc_on_unimplemented(
-    note = "consider using `Box::pin`",
+    note = "consider using `core::pin::pin!`\nconsider using `Box::pin` if you want ownership of the pinned value",
```
Comment (cjgillot) [inline library/core/src/marker.rs:826]: What's the policy about mentioning `core` in an diagnostic message? Should we mention the `pin!` macro instead?

I'm not convinced about the "if you want ownership of the pinned value" part. The important difference between the two is the local storage. What about: "if you need to access the pinned value outside of the current scope"?

Comment (ndrewxie) [reply library/core/src/marker.rs:826]: Yeah, I suppose the std::pin::pin! macro would be better

Also yeah, that message works better, it captures the "local pin" semantics far better. I'll make the changes and commit

File: library/core/src/marker.rs [hunk 2]
```diff
@@ -823,7 +823,7 @@ unsafe impl<T: ?Sized> Freeze for &mut T {}
 /// [`pin` module]: crate::pin
 #[stable(feature = "pin", since = "1.33.0")]
 #[rustc_on_unimplemented(
-    note = "consider using `Box::pin`",
+    note = "consider using `std::pin::pin!`\nconsider using `Box::pin` if you need to access the pinned value outside of the current scope",
```
Comment (cjgillot) [inline library/core/src/marker.rs:826]:
```suggestion
    note = "consider using the `pin!` macro\nconsider using `Box::pin` if you need to access the pinned value outside of the current scope",
```
Works for both std and no-std :smile:

r=me with commits squashed

**Comment by ndrewxie** [comment]
Whoops, I was trying to sort out some git stuff (git won't let me switch to the branch for this issue for some reason, git clone literally didn't clone it), and in the process of renaming the branch I accidentally closed it... Is it OK for me to create a new PR and r= you, @cjgillot ? Sorry😅

**Pull request closed without merging**